traffic_router/core/src/main/java/com/comcast/cdn/traffic_control/traffic_router/core/router/TrafficRouter.java

```diff
@@ -560,16 +564,50 @@ public HTTPRouteResult multiRoute(final HTTPRequest request, final Track track)
 		return routeResult;
 	}
 
+	/**
+	 * Creates a string to be used in consistent hashing
+	 *
+	 * This uses simply the request path by default, but will consider any and all Query Parameters
+	 * that are in deliveryService's {@link DeliveryService.consistentHashQueryParams} set as well.
+	 * It will also fall back on the request path if the query parameters are not UTF-8-encoded.
+	 *
+	 * @param deliveryService The Delivery Service being requested
+	 * @param request An {@link HTTPRequest} representing the client's request.
+	 * @return A string appropriate to use for consistent hashing to service the request
+	*/
+	@SuppressWarnings({"PMD.CyclomaticComplexity", "PMD.UseStringBufferForStringAppends"})
 	public String buildPatternBasedHashString(final DeliveryService deliveryService, final HTTPRequest request) {
 		final String requestPath = request.getPath();
 		if (deliveryService.getConsistentHashRegex() != null && !deliveryService.getConsistentHashRegex().isEmpty() && !requestPath.isEmpty()) {
 			final StringBuilder hashString = new StringBuilder(request.getPath());
-			for (String qkey : request.getQueryString().split("&")) {
-				qkey = qkey.split("=")[0];
-				if (deliveryService.consistentHashQueryParams.contains(qkey)) {
-					hashString.append(qkey);
+			final SortedSet<String> qparams = new TreeSet<String>();
+			try {
+				for (final String qparam : request.getQueryString().split("&")) {
+					if (qparam.isEmpty()) {
+						continue;
+					}
+
+					String[] parts = qparam.split("=");
+					for (short i = 0; i<parts.length; ++i) {
+						parts[i] = URLDecoder.decode(parts[i]);
+					}
+					parts[0] = parts[0].toLowerCase(); // I truly have no idea why this triggers `PMD.UseStringBufferForStringAppends`
+
+					if (deliveryService.getConsistentHashQueryParams().contains(parts[0])) {
+						if (parts.length == 2) {
+							parts[1] = URLDecoder.decode(parts[1]);
+						}
+						qparams.add(URLEncoder.encode(String.join("=", parts)));
```

Comment:
Why are we re-encoding before appending to the hashString? I thought we wanted to hash on the decoded qparams.

Reply:
The point was to compare things in an encoding-consistent way. But I guess that's accomplished by just doing the decoding.